recipes-bsp/imx-mkimage/imx-boot_0.2.bb

```diff
@@ -66,6 +68,8 @@ SOC_TARGET_mx8qxp = "iMX8QX"
 SOC_TARGET_mx8mq  = "iMX8M"
 SOC_TARGET_mx8mm  = "iMX8MM"
 
+DEPLOY_OPTEE = "${@bb.utils.contains('COMBINED_FEATURES', 'optee', 'true', 'false', d)}"
+
```

Comment:
Not need. You can check if the file exists and copy it if it does.

Reply:
Consider the negative case where optee is expected but missing. It will be better to have the recipe fail instead of being silent, right?

Reply:
We use a distro feature here to allow the user to decide if he wants optee. This is the point of COMBINED_FEATURES. It is perfectly valid for the machine to support optee but the user to disable it.
My point is about the case where the machine supports optee, and the user wants it, but the file is missing. This should cause the recipe to fail.

Reply:
This is not the use case of a distro feature. Instead, user can do:

```
MACHINE_FEATURES_append = " optee"
```

and enable it when desired.

Reply:
That's not what the mega-manual says. Here's the MACHINE_FEATURES definition, emphasis mine: Specifies the list of hardware features the MACHINE is _capable_ of supporting. For related information on _enabling_ features, see the DISTRO_FEATURES, COMBINED_FEATURES, and IMAGE_FEATURES variables.

Reply:
And from the COMBINED_FEATURES definition: For example, the "bluetooth" feature requires hardware support but should also be optional at the distribution level, in case the hardware supports Bluetooth but you do not ever intend to use it.

Reply:
Yes but this is intended for use with things which has to enable support in multiple components as for example OpenGL, ALSA or as you mentioned Bluetooth. 

All those need to change all the system and add/remove build flags and dependencies. The `optee` specifically is a machine-only change and I will not pollute the `DISTRO_FEATURES` namespace with something which is i.MX specific.

The possible use of it inside the `MACHINE_FEATURES` allows for easy addition of it while leaves the decision to the user.

Reply:
imx-atf, imx-boot, and image_types_fsl.bbclass (which I haven't pushed yet) are the multiple components that need to enable support for optee.

Your suggestion uses MACHINE_FEATURES in an undocumented way. Is your use model missing from the manual?

As it is documented, MACHINE_FEATURES is an invariant. The MACHINE supports a feature or it doesn't. If the user needs to be able to enable it, then DISTRO_FEATURES and COMBINED_FEATURES come into play. Seems like a nice, clear use model. And this is what is documented. What is wrong with it?

Reply:
Yes, all those components are meta-freescale ones. A distribution goes beyond the i.MX. If I am using an Intel or a Marvell machine, does the `optee` makes sense? of course not. So it is not a `DISTRO_FEATURE` but a `BSP_FEATURE` which does not exists. The closest is `MACHINE_FEATURES`.

Reply:
But I think what's important is that optee itself is not an i.MX-specific concept. Others could use it to control their particular implementation as well, whatever the distro, whatever the machine.

For Marvell and Intel, the optee distro feature would be fine because it just says that the user would like it if available. Without the machine support of course the image won't have it. Not only does it make sense, it's the main point of COMBINED_FEATURES.

Reply:
Sure, but then it must be sent to OE-Core as this is where the generic distro features belongs.

Reply:
Okay, I'm still not convinced, but I'm not going to argue any more. I will remove the usage of COMBINED_FEATURES.